src/firmware_new/tools/communication/transfer_media/network_medium.cpp

```diff
@@ -0,0 +1,96 @@
+#include "firmware_new/tools/communication/transfer_media/network_medium.h"
+
+#include <g3log/g3log.hpp>
+
+#include "boost/array.hpp"
+#include "boost/asio.hpp"
+#include "boost/asio/error.hpp"
+#include "boost/bind.hpp"
+
+using boost::asio::socket_base;
+using boost::asio::ip::address_v4;
+using boost::asio::ip::udp;
+
+NetworkMedium::NetworkMedium(const std::string& local_ipaddr, unsigned port)
+{
+    socket.reset(new udp::socket(io_service));
+
+    // TODO change to multicast after firmware is on RTOS
+    local_endpoint     = udp::endpoint(address_v4::any(), port);
+    broadcast_endpoint = udp::endpoint(address_v4::from_string(local_ipaddr), port);
+
+    socket->open(local_endpoint.protocol());
+
+    try
+    {
+        socket->bind(local_endpoint);
+    }
+    catch (const boost::exception& ex)
+    {
+        LOG(WARNING) << "There was an issue binding the socket to the endpoint when"
+                        "trying to connect to the provided port"
+                        "Please make sure no other program is using the port"
+                     << std::endl;
+
+        // Throw this exception up to top-level, as we have no valid
+        // recovery action here
+        throw;
```

Comment:
I think you want `std::rethrow_exception` (http://www.cplusplus.com/reference/exception/rethrow_exception/) here?

Reply:
https://stackoverflow.com/questions/19984686/boost-rethrow-exception-throws-exception-of-different-type

looks like boost doesn't let you do this by design, `throw` does rethrow the same exception